sql: creating foreign keys between two single-col (pk) tables times out

Adding these tests to `sql/testdata/fk` causes a timeout:

```
statement ok
CREATE DATABASE test2

statement ok
CREATE TABLE test2.other(x INT PRIMARY KEY)

statement ok
CREATE TABLE crossdb (y INT PRIMARY KEY, FOREIGN KEY (y) REFERENCES test2.other(x))

statement error foreign key violation: value \[2\] not found in other@primary \[x\]
INSERT INTO crossdb(y) VALUES (2)
```

```
CREATE TABLE crossdb (y INT PRIMARY KEY, FOREIGN KEY (y) REFERENCES test2.other(x));I160808 18:08:55.988937 storage/replica_command.go:2180  store=1:1 range=19 [/Table/63-/Max): initiating a split of this range at key /Table/64
I160808 18:08:55.989643 storage/log.go:84  Range Event: "split", range: 19, info: {"UpdatedDesc":{"range_id":19,"start_key":"xw==","end_key":"yA==","replicas":[{"node_id":1,"store_id":1,"replica_id":1}],"next_replica_id":2},"NewDesc":{"range_id":20,"start_key":"yA==","end_key":"//8=","replicas":[{"node_id":1,"store_id":1,"replica_id":1}],"next_replica_id":2}}
W160808 18:08:55.991655 storage/intent_resolver.go:344  could not GC completed transaction: end key /Local/Range/"\xc7"/"rdsc\x00" must be greater than start /Local/Range/"\xc7"/RangeDescriptor
I160808 18:09:04.816644 server/status/runtime.go:189  runtime stats: 50 MiB RSS, 55 goroutines, 10 MiB/2.6 MiB/18 MiB GO alloc/idle/total, 19 MiB/25 MiB CGO alloc/total, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (135x)
...
panic: test timed out after 2m0s
```

I was adding these as part of #8152. I will comment them out and reference this issue.

--- a/sql/backfill.go
+++ b/sql/backfill.go
@@ -270,11 +270,13 @@ func (sc *SchemaChanger) truncateAndBackfillColumnsChunk(
 		}
 
 		updateCols := append(added, dropped...)
-		fkTables := TablesNeededForFKs(*tableDesc, CheckUpdates)
+		fkTables := tablesNeededForFKs(*tableDesc, CheckUpdates)
 		for k := range fkTables {
-			if fkTables[k], err = sqlbase.GetTableDescFromID(txn, k); err != nil {
+			table, err := sqlbase.GetTableDescFromID(txn, k)
+			if err != nil {
 				return err
 			}
+			fkTables[k] = tableLookup{table: table}
 		}
 		// TODO(dan): Tighten up the bound on the requestedCols parameter to
 		// makeRowUpdater.

--- a/sql/create.go
+++ b/sql/create.go
@@ -349,7 +349,10 @@ func (n *createTableNode) Start() error {
 	}
 
 	// FKs are resolved after the descriptor is otherwise complete and IDs have
-	// been allocated since the FKs will reference those IDs.
+	// been allocated since the FKs will reference those IDs. Resolution also
+	// accumulated updates to other tables (adding backreferences) in the passed
+	// map -- anything in that map should be saved when the table is created.
+	affected := make(map[sqlbase.ID]*sqlbase.TableDescriptor)
 	for _, def := range n.n.Defs {
 		switch d := def.(type) {
 		case *parser.ColumnTableDef:
@@ -359,13 +362,13 @@ func (n *createTableNode) Start() error {
 					targetCol = append(targetCol, d.References.Col)
 				}
 				err := n.resolveFK(&desc, parser.NameList{d.Name},
-					d.References.Table.TableName(), targetCol, d.References.ConstraintName)
+					d.References.Table.TableName(), targetCol, d.References.ConstraintName, affected)
 				if err != nil {
 					return err
 				}
 			}
 		case *parser.ForeignKeyConstraintTableDef:
-			err := n.resolveFK(&desc, d.FromCols, d.Table.TableName(), d.ToCols, d.Name)
+			err := n.resolveFK(&desc, d.FromCols, d.Table.TableName(), d.ToCols, d.Name, affected)
 			if err != nil {
 				return err
 			}
@@ -400,8 +403,13 @@ func (n *createTableNode) Start() error {
 	}
 
 	if created {
-		if err := n.finalizeFKs(&desc); err != nil {
-			return err
+		for _, updated := range affected {
+			if err := n.p.saveNonmutationAndNotify(updated); err != nil {
+				return err
+			}
+		}
+		if desc.Adding() {
+			n.p.notifySchemaChange(desc.ID, sqlbase.InvalidMutationID)
 		}
 
 		for _, index := range desc.AllNonDropIndexes() {
@@ -447,28 +455,52 @@ func (n *createTableNode) ExplainPlan(v bool) (string, string, []planNode) {
 	return "create table", "", nil
 }
 
+// resolveFK looks up the tables and columns mentioned in a `REFERENCES`
+// constraint and adds metadata representing that constraint to the descriptor.
+// It may, in doing so, add to or alter descriptors in the passed in `backrefs`
+// map of other tables that need to be updated when this table is created.
 func (n *createTableNode) resolveFK(
 	tbl *sqlbase.TableDescriptor,
-	fromCols parser.NameList,
+	fromColNames parser.NameList,
 	targetTable *parser.TableName,
 	targetColNames parser.NameList,
 	constraintName parser.Name,
+	backrefs map[sqlbase.ID]*sqlbase.TableDescriptor,
 ) error {
-	srcCols, err := tbl.FindActiveColumnsByNames(fromCols)
-	if err != nil {
-		return err
-	}
-
 	target, err := n.p.getTableDesc(targetTable)
 	if err != nil {
 		return err
 	}
+	// Special-case: self-referencing FKs (i.e. referencing another col in the
+	// same table) will reference a table name that doesn't exist yet (since we
+	// are creating it).
 	if target == nil {
 		if targetTable.Table() == tbl.Name {
 			target = tbl
 		} else {
 			return fmt.Errorf("referenced table %q not found", targetTable.String())
 		}
+	} else {
+		// Since this FK is referencing another table, this table must be created in
+		// a non-public "ADD" state and made public only after all leases on the
+		// other table are updated to include the backref.
+		tbl.State = sqlbase.TableDescriptor_ADD
+		if err := tbl.SetUpVersion(); err != nil {
+			return err
+		}
+
+		// If we resolve the same table more than once, we only want to edit a
+		// single instance of it, so replace target with previously resolved table.
+		if prev, ok := backrefs[target.ID]; ok {
+			target = prev
+		} else {
+			backrefs[target.ID] = target
+		}
+	}
+
+	srcCols, err := tbl.FindActiveColumnsByNames(fromColNames)
+	if err != nil {
+		return err
 	}
 
 	// If no columns are specified, attempt to default to PK.
@@ -521,18 +553,18 @@ func (n *createTableNode) resolveFK(
 	}
 
 	if constraintName == "" {
-		constraintName = parser.Name(fmt.Sprintf("fk_%s_ref_%s", fromCols[0], target.Name))
+		constraintName = parser.Name(fmt.Sprintf("fk_%s_ref_%s", fromColNames[0], target.Name))
 	}
 
-	var targetIdx sqlbase.IndexID
+	var targetIdx *sqlbase.IndexDescriptor
 	if matchesIndex(targetCols, target.PrimaryIndex, matchExact) {
-		targetIdx = target.PrimaryIndex.ID
+		targetIdx = &target.PrimaryIndex
 	} else {
 		found := false
 		// Find the index corresponding to the referenced column.
-		for _, idx := range target.Indexes {
+		for i, idx := range target.Indexes {
 			if idx.Unique && matchesIndex(targetCols, idx, matchExact) {
-				targetIdx = idx.ID
+				targetIdx = &target.Indexes[i]
 				found = true
 				break
 			}
@@ -542,49 +574,30 @@ func (n *createTableNode) resolveFK(
 		}
 	}
 
-	ref := sqlbase.ForeignKeyReference{Table: target.ID, Index: targetIdx, Name: string(constraintName)}
+	ref := sqlbase.ForeignKeyReference{Table: target.ID, Index: targetIdx.ID, Name: string(constraintName)}
+	backref := sqlbase.ForeignKeyReference{Table: tbl.ID}
 
-	var srcIdx sqlbase.IndexID
 	if matchesIndex(srcCols, tbl.PrimaryIndex, matchPrefix) {
 		if tbl.PrimaryIndex.ForeignKey.IsSet() {
 			return fmt.Errorf("column(s) cannot be used by multiple foreign key constraints")
 		}
 		tbl.PrimaryIndex.ForeignKey = ref
-		srcIdx = tbl.PrimaryIndex.ID
-	} else {
-		found := false
-		for i := range tbl.Indexes {
-			if matchesIndex(srcCols, tbl.Indexes[i], matchPrefix) {
-				if tbl.Indexes[i].ForeignKey.IsSet() {
-					return fmt.Errorf("column(s) cannot be used by multiple foreign key constraints")
-				}
-				tbl.Indexes[i].ForeignKey = ref
-				srcIdx = tbl.Indexes[i].ID
-				found = true
-				break
+		backref.Index = tbl.PrimaryIndex.ID
+		targetIdx.ReferencedBy = append(targetIdx.ReferencedBy, backref)
+		return nil
+	}
+	for i := range tbl.Indexes {
+		if matchesIndex(srcCols, tbl.Indexes[i], matchPrefix) {
+			if tbl.Indexes[i].ForeignKey.IsSet() {
+				return fmt.Errorf("column(s) cannot be used by multiple foreign key constraints")
 			}
-		}
-		if !found {
-			return fmt.Errorf("foreign key columns %s must be the prefix of an index", colNames(srcCols))
+			tbl.Indexes[i].ForeignKey = ref
+			backref.Index = tbl.Indexes[i].ID
+			targetIdx.ReferencedBy = append(targetIdx.ReferencedBy, backref)
+			return nil
 		}
 	}
-
-	// self-reference FKs can setup their back-references at creation without
-	// waiting for the finalize step, and do not need to use the ADD table state.
-	if target == tbl {
-		ref.Name = ""
-		ref.Index = srcIdx
-		if idx, err := tbl.FindIndexByID(targetIdx); err == nil {
-			idx.ReferencedBy = append(idx.ReferencedBy, ref)
-		} else {
-			return err
-		}
-	} else {
-		// Create the table non-public, since we'll need to ensure the FK back-refs
-		// are in place before we can allow writes.
-		tbl.State = sqlbase.TableDescriptor_ADD
-	}
-	return nil
+	return fmt.Errorf("foreign key columns %s must be the prefix of an index", colNames(srcCols))
 }
 
 // colNames converts a []colDesc to a human-readable string for use in error messages.
@@ -674,47 +687,6 @@ func (p *planner) addInterleave(
 	index.Interleave = sqlbase.InterleaveDescriptor{Ancestors: append(ancestorPrefix, intl)}
 
 	desc.State = sqlbase.TableDescriptor_ADD
-	return nil
-}
-
-func (n *createTableNode) finalizeFKs(desc *sqlbase.TableDescriptor) error {
-	save := make(map[sqlbase.ID]*sqlbase.TableDescriptor)
-	for _, idx := range desc.Indexes {
-		if idx.ForeignKey.IsSet() {
-			if idx.ForeignKey.Table == desc.ID {
-				// self-ref FKs include back-references at creation.
-				continue
-			}
-			targetTable, ok := save[idx.ForeignKey.Table]
-			if !ok {
-				var err error
-				targetTable, err = sqlbase.GetTableDescFromID(n.p.txn, idx.ForeignKey.Table)
-				if err != nil {
-					return err
-				}
-				save[targetTable.ID] = targetTable
-			}
-			targetIdx, err := targetTable.FindIndexByID(idx.ForeignKey.Index)
-			if err != nil {
-				return err
-			}
-			backref := sqlbase.ForeignKeyReference{Table: desc.ID, Index: idx.ID}
-			targetIdx.ReferencedBy = append(targetIdx.ReferencedBy, backref)
-		}
-	}
-	for _, t := range save {
-		if err := n.p.saveNonmutationAndNotify(t); err != nil {
-			return err
-		}
-	}
-
-	if desc.State == sqlbase.TableDescriptor_ADD {
-		desc.State = sqlbase.TableDescriptor_PUBLIC
-
-		if err := n.p.saveNonmutationAndNotify(desc); err != nil {
-			return err
-		}
-	}
 	return nil
 }
 

--- a/sql/delete.go
+++ b/sql/delete.go
@@ -64,7 +64,7 @@ func (p *planner) Delete(n *parser.Delete, desiredTypes []parser.Datum, autoComm
 		requestedCols = en.tableDesc.Columns
 	}
 
-	fkTables := TablesNeededForFKs(*en.tableDesc, CheckDeletes)
+	fkTables := tablesNeededForFKs(*en.tableDesc, CheckDeletes)
 	if err := p.fillFKTableMap(fkTables); err != nil {
 		return nil, err
 	}

--- a/sql/descriptor_mutation_test.go
+++ b/sql/descriptor_mutation_test.go
@@ -948,3 +948,49 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 		}
 	}
 }
+
+// TestAddingFKs checks the behavior of a table in the non-public `ADD` state.
+// Being non-public, it should not be visible to clients, and is therefore
+// assumed to be empty (e.g. by foreign key checks), since no one could have
+// written to it yet.
+func TestAddingFKs(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	params, _ := createTestServerParams()
+	s, sqlDB, kvDB := serverutils.StartServer(t, params)
+	defer s.Stopper().Stop()
+
+	if _, err := sqlDB.Exec(`
+		CREATE DATABASE t;
+		CREATE TABLE t.products (id INT PRIMARY KEY);
+		INSERT INTO t.products VALUES (1), (2);
+		CREATE TABLE t.orders (id INT PRIMARY KEY, product INT REFERENCES t.products, INDEX (product));
+	`); err != nil {
+		t.Fatal(err)
+	}
+
+	// Step the referencing table back to the ADD state.
+	ordersDesc := sqlbase.GetTableDescriptor(kvDB, "t", "orders")
+	ordersDesc.State = sqlbase.TableDescriptor_ADD
+	ordersDesc.Version++
+	if err := kvDB.Put(
+		sqlbase.MakeDescMetadataKey(ordersDesc.ID),
+		sqlbase.WrapDescriptor(ordersDesc),
+	); err != nil {
+		t.Fatal(err)
+	}
+
+	// Generally a referenced table needs to lookup referencing tables to check
+	// FKs during delete operations, but referencing tables in the ADD state are
+	// given special treatment.
+	if _, err := sqlDB.Exec(`DELETE FROM t.products`); err != nil {
+		t.Fatal(err)
+	}
+
+	// Client should not see the orders table.
+	if _, err := sqlDB.Exec(
+		`SELECT * FROM t.orders`,
+	); !testutils.IsError(err, "table is being added") {
+		t.Fatal(err)
+	}
+}

--- a/sql/fk.go
+++ b/sql/fk.go
@@ -24,8 +24,13 @@ import (
 	"github.com/pkg/errors"
 )
 
-// TablesByID maps table IDs to looked up descriptors.
-type TablesByID map[sqlbase.ID]*sqlbase.TableDescriptor
+// tableLookupsByID maps table IDs to looked up descriptors or, for tables that
+// exist but are not yet public/leasable, entries with just the isAdding flag.
+type tableLookupsByID map[sqlbase.ID]tableLookup
+type tableLookup struct {
+	table    *sqlbase.TableDescriptor
+	isAdding bool
+}
 
 // FKCheck indicates a kind of FK check (delete, insert, or both).
 type FKCheck int
@@ -39,29 +44,29 @@ const (
 	CheckUpdates
 )
 
-// TablesNeededForFKs calculates the IDs of the additional TableDescriptors that
+// tablesNeededForFKs calculates the IDs of the additional TableDescriptors that
 // will be needed for FK checking delete and/or insert operations on `table`.
 //
 // NB: the returned map's values are *not* set -- higher level calling code, eg
 // in planner, should fill the map's values by acquiring leases. This function
 // is essentially just returning a slice of IDs, but the empty map can be filled
 // in place and reused, avoiding a second allocation.
-func TablesNeededForFKs(table sqlbase.TableDescriptor, usage FKCheck) TablesByID {
-	var ret TablesByID
+func tablesNeededForFKs(table sqlbase.TableDescriptor, usage FKCheck) tableLookupsByID {
+	var ret tableLookupsByID
 	for _, idx := range table.AllNonDropIndexes() {
 		if usage != CheckDeletes && idx.ForeignKey.IsSet() {
 			if ret == nil {
-				ret = make(TablesByID)
+				ret = make(tableLookupsByID)
 			}
-			ret[idx.ForeignKey.Table] = nil
+			ret[idx.ForeignKey.Table] = tableLookup{}
 		}
 		if usage != CheckInserts {
 			for _, idx := range table.AllNonDropIndexes() {
 				for _, ref := range idx.ReferencedBy {
 					if ret == nil {
-						ret = make(TablesByID)
+						ret = make(tableLookupsByID)
 					}
-					ret[ref.Table] = nil
+					ret[ref.Table] = tableLookup{}
 				}
 			}
 		}
@@ -74,7 +79,7 @@ type fkInsertHelper map[sqlbase.IndexID][]baseFKHelper
 var errSkipUnsedFK = errors.New("no columns involved in FK included in writer")
 
 func makeFKInsertHelper(
-	txn *client.Txn, table sqlbase.TableDescriptor, otherTables TablesByID, colMap map[sqlbase.ColumnID]int,
+	txn *client.Txn, table sqlbase.TableDescriptor, otherTables tableLookupsByID, colMap map[sqlbase.ColumnID]int,
 ) (fkInsertHelper, error) {
 	var fks fkInsertHelper
 	for _, idx := range table.AllNonDropIndexes() {
@@ -136,11 +141,16 @@ func (fks fkInsertHelper) checkIdx(idx sqlbase.IndexID, row parser.DTuple) error
 type fkDeleteHelper map[sqlbase.IndexID][]baseFKHelper
 
 func makeFKDeleteHelper(
-	txn *client.Txn, table sqlbase.TableDescriptor, otherTables TablesByID, colMap map[sqlbase.ColumnID]int,
+	txn *client.Txn, table sqlbase.TableDescriptor, otherTables tableLookupsByID, colMap map[sqlbase.ColumnID]int,
 ) (fkDeleteHelper, error) {
 	var fks fkDeleteHelper
 	for _, idx := range table.AllNonDropIndexes() {
 		for _, ref := range idx.ReferencedBy {
+			if otherTables[ref.Table].isAdding {
+				// We can assume that a table being added but not yet public is empty,
+				// and thus does not need to be checked for FK violations.
+				continue
+			}
 			fk, err := makeBaseFKHelper(txn, otherTables, idx, ref, colMap)
 			if err == errSkipUnsedFK {
 				continue
@@ -195,7 +205,7 @@ type fkUpdateHelper struct {
 }
 
 func makeFKUpdateHelper(
-	txn *client.Txn, table sqlbase.TableDescriptor, otherTables TablesByID, colMap map[sqlbase.ColumnID]int,
+	txn *client.Txn, table sqlbase.TableDescriptor, otherTables tableLookupsByID, colMap map[sqlbase.ColumnID]int,
 ) (fkUpdateHelper, error) {
 	ret := fkUpdateHelper{}
 	var err error
@@ -226,19 +236,17 @@ type baseFKHelper struct {
 
 func makeBaseFKHelper(
 	txn *client.Txn,
-	otherTables TablesByID,
+	otherTables tableLookupsByID,
 	writeIdx sqlbase.IndexDescriptor,
 	ref sqlbase.ForeignKeyReference,
 	colMap map[sqlbase.ColumnID]int, // col ids (for idx being written) to row offset.
 ) (baseFKHelper, error) {
-	b := baseFKHelper{txn: txn, writeIdx: writeIdx}
-	searchTable, ok := otherTables[ref.Table]
-	if !ok {
+	b := baseFKHelper{txn: txn, writeIdx: writeIdx, searchTable: otherTables[ref.Table].table}
+	if b.searchTable == nil {
 		return b, errors.Errorf("referenced table %d not in provided table map %+v", ref.Table, otherTables)
 	}
-	b.searchTable = searchTable
 	b.searchPrefix = sqlbase.MakeIndexKeyPrefix(b.searchTable, ref.Index)
-	searchIdx, err := searchTable.FindIndexByID(ref.Index)
+	searchIdx, err := b.searchTable.FindIndexByID(ref.Index)
 	if err != nil {
 		return b, err
 	}
@@ -247,13 +255,13 @@ func makeBaseFKHelper(
 		b.prefixLen = len(writeIdx.ColumnIDs)
 	}
 	b.searchIdx = searchIdx
-	ids := colIDtoRowIndexFromCols(searchTable.Columns)
+	ids := colIDtoRowIndexFromCols(b.searchTable.Columns)
 	needed := make([]bool, len(ids))
 	for _, i := range searchIdx.ColumnIDs {
 		needed[ids[i]] = true
 	}
-	isSecondary := searchTable.PrimaryIndex.ID != searchIdx.ID
-	err = b.rf.Init(searchTable, ids, searchIdx, false, isSecondary, searchTable.Columns, needed)
+	isSecondary := b.searchTable.PrimaryIndex.ID != searchIdx.ID
+	err = b.rf.Init(b.searchTable, ids, searchIdx, false, isSecondary, b.searchTable.Columns, needed)
 	if err != nil {
 		return b, err
 	}

--- a/sql/insert.go
+++ b/sql/insert.go
@@ -150,7 +150,7 @@ func (p *planner) Insert(
 		return nil, fmt.Errorf("INSERT error: table %s has %d columns but %d values were supplied", n.Table, numInputColumns, expressions)
 	}
 
-	fkTables := TablesNeededForFKs(*en.tableDesc, CheckInserts)
+	fkTables := tablesNeededForFKs(*en.tableDesc, CheckInserts)
 	if err := p.fillFKTableMap(fkTables); err != nil {
 		return nil, err
 	}
@@ -188,7 +188,7 @@ func (p *planner) Insert(
 				return nil, err
 			}
 
-			fkTables := TablesNeededForFKs(*en.tableDesc, CheckUpdates)
+			fkTables := tablesNeededForFKs(*en.tableDesc, CheckUpdates)
 			if err := p.fillFKTableMap(fkTables); err != nil {
 				return nil, err
 			}

--- a/sql/lease.go
+++ b/sql/lease.go
@@ -124,8 +124,6 @@ func jitteredLeaseDuration() time.Duration {
 	return time.Duration(float64(LeaseDuration) * (0.75 + 0.5*rand.Float64()))
 }
 
-var errTableDeleted = errors.New("table is being deleted")
-
 // Acquire a lease on the most recent version of a table descriptor.
 // If the lease cannot be obtained because the descriptor is in the process of
 // being deleted, the error will be errTableDeleted.
@@ -160,13 +158,12 @@ func (s LeaseStore) Acquire(
 	}
 
 	tableDesc := desc.GetTable()
-	if tableDesc != nil && tableDesc.Deleted() {
-		return nil, errTableDeleted
-	}
-	if tableDesc == nil || tableDesc.State != sqlbase.TableDescriptor_PUBLIC {
+	if tableDesc == nil {
 		return nil, errors.Errorf("ID %d is not a table", tableID)
 	}
-
+	if err := filterTableState(tableDesc); err != nil {
+		return nil, err
+	}
 	tableDesc.MaybeUpgradeFormatVersion()
 	lease.TableDescriptor = *tableDesc
 

--- a/sql/planner.go
+++ b/sql/planner.go
@@ -300,12 +300,17 @@ func (p *planner) checkTestingVerifyMetadataOrDie(
 	p.testingVerifyMetadataFn = nil
 }
 
-func (p *planner) fillFKTableMap(m TablesByID) error {
-	var err error
+func (p *planner) fillFKTableMap(m tableLookupsByID) error {
 	for tableID := range m {
-		if m[tableID], err = p.getTableLeaseByID(tableID); err != nil {
+		table, err := p.getTableLeaseByID(tableID)
+		if err == errTableAdding {
+			m[tableID] = tableLookup{isAdding: true}
+			continue
+		}
+		if err != nil {
 			return err
 		}
+		m[tableID] = tableLookup{table: table}
 	}
 	return nil
 }

--- a/sql/rowwriter.go
+++ b/sql/rowwriter.go
@@ -150,7 +150,7 @@ type rowInserter struct {
 func makeRowInserter(
 	txn *client.Txn,
 	tableDesc *sqlbase.TableDescriptor,
-	fkTables TablesByID,
+	fkTables tableLookupsByID,
 	insertCols []sqlbase.ColumnDescriptor,
 	checkFKs bool,
 ) (rowInserter, error) {
@@ -372,7 +372,7 @@ const (
 func makeRowUpdater(
 	txn *client.Txn,
 	tableDesc *sqlbase.TableDescriptor,
-	fkTables TablesByID,
+	fkTables tableLookupsByID,
 	updateCols []sqlbase.ColumnDescriptor,
 	requestedCols []sqlbase.ColumnDescriptor,
 	updateType rowUpdaterType,
@@ -754,7 +754,7 @@ type rowDeleter struct {
 func makeRowDeleter(
 	txn *client.Txn,
 	tableDesc *sqlbase.TableDescriptor,
-	fkTables TablesByID,
+	fkTables tableLookupsByID,
 	requestedCols []sqlbase.ColumnDescriptor,
 	checkFKs bool,
 ) (rowDeleter, error) {
@@ -800,7 +800,7 @@ func makeRowDeleter(
 	if checkFKs {
 		var err error
 		if rd.fks, err = makeFKDeleteHelper(txn, *tableDesc, fkTables, fetchColIDtoRowIndex); err != nil {
-			return rowDeleter{}, nil
+			return rowDeleter{}, err
 		}
 	}
 

--- a/sql/schema_changer.go
+++ b/sql/schema_changer.go
@@ -222,8 +222,9 @@ func (sc SchemaChanger) exec(
 	if err != nil {
 		return err
 	}
+	table := desc.GetTable()
 
-	if desc.GetTable().Deleted() {
+	if table.Deleted() {
 		lease, err = sc.ExtendLease(lease)
 		if err != nil {
 			return err
@@ -231,19 +232,19 @@ func (sc SchemaChanger) exec(
 		// Wait for everybody to see the version with the deleted bit set. When
 		// this returns, nobody has any leases on the table, nor can get new leases,
 		// so the table will no longer be modified.
-		if err := sc.waitToUpdateLeases(); err != nil {
+		if err := sc.waitToUpdateLeases(sc.tableID); err != nil {
 			return err
 		}
 
 		// Truncate the table and delete the descriptor.
-		if err := sc.truncateAndDropTable(&lease, desc.GetTable()); err != nil {
+		if err := sc.truncateAndDropTable(&lease, table); err != nil {
 			return err
 		}
 		needRelease = false
 		return nil
 	}
 
-	if desc.GetTable().Renamed() {
+	if table.Renamed() {
 		lease, err = sc.ExtendLease(lease)
 		if err != nil {
 			return err
@@ -251,7 +252,7 @@ func (sc SchemaChanger) exec(
 		// Wait for everyone to see the version with the new name. When this
 		// returns, no new transactions will be using the old name for the table, so
 		// the old name can now be re-used (by CREATE).
-		if err := sc.waitToUpdateLeases(); err != nil {
+		if err := sc.waitToUpdateLeases(sc.tableID); err != nil {
 			return err
 		}
 
@@ -261,7 +262,7 @@ func (sc SchemaChanger) exec(
 		// Free up the old name(s).
 		err := sc.db.Txn(func(txn *client.Txn) error {
 			b := txn.NewBatch()
-			for _, renameDetails := range desc.GetTable().Renames {
+			for _, renameDetails := range table.Renames {
 				tbKey := tableKey{renameDetails.OldParentID, renameDetails.OldName}.Key()
 				b.Del(tbKey)
 			}
@@ -285,7 +286,7 @@ func (sc SchemaChanger) exec(
 	// returns, so that the new schema is live everywhere. This is not needed for
 	// correctness but is done to make the UI experience/tests predictable.
 	defer func() {
-		if err := sc.waitToUpdateLeases(); err != nil {
+		if err := sc.waitToUpdateLeases(sc.tableID); err != nil {
 			log.Warning(context.TODO(), err)
 		}
 	}()
@@ -396,12 +397,12 @@ func (sc *SchemaChanger) RunStateMachineBeforeBackfill() error {
 		return err
 	}
 	// wait for the state change to propagate to all leases.
-	return sc.waitToUpdateLeases()
+	return sc.waitToUpdateLeases(sc.tableID)
 }
 
 // Wait until the entire cluster has been updated to the latest version
 // of the table descriptor.
-func (sc *SchemaChanger) waitToUpdateLeases() error {
+func (sc *SchemaChanger) waitToUpdateLeases(tableID sqlbase.ID) error {
 	// Aggressively retry because there might be a user waiting for the
 	// schema change to complete.
 	retryOpts := retry.Options{
@@ -410,11 +411,11 @@ func (sc *SchemaChanger) waitToUpdateLeases() error {
 		Multiplier:     2,
 	}
 	if log.V(2) {
-		log.Infof(context.TODO(), "waiting for a single version of table %d...", sc.tableID)
+		log.Infof(context.TODO(), "waiting for a single version of table %d...", tableID)
 	}
-	_, err := sc.leaseMgr.waitForOneVersion(sc.tableID, retryOpts)
+	_, err := sc.leaseMgr.waitForOneVersion(tableID, retryOpts)
 	if log.V(2) {
-		log.Infof(context.TODO(), "waiting for a single version of table %d... done", sc.tableID)
+		log.Infof(context.TODO(), "waiting for a single version of table %d... done", tableID)
 	}
 	return err
 }

--- a/sql/schema_changer.go
+++ b/sql/schema_changer.go
@@ -244,6 +244,27 @@ func (sc SchemaChanger) exec(
 		return nil
 	}
 
+	if table.Adding() {
+		for _, idx := range table.AllNonDropIndexes() {
+			if idx.ForeignKey.IsSet() {
+				if err := sc.waitToUpdateLeases(idx.ForeignKey.Table); err != nil {
+					return err
+				}
+			}
+		}
+
+		if _, err := sc.leaseMgr.Publish(
+			table.ID,
+			func(tbl *sqlbase.TableDescriptor) error {
+				tbl.State = sqlbase.TableDescriptor_PUBLIC
+				return nil
+			},
+			func(txn *client.Txn) error { return nil },
+		); err != nil {
+			return err
+		}
+	}
+
 	if table.Renamed() {
 		lease, err = sc.ExtendLease(lease)
 		if err != nil {
@@ -724,7 +745,7 @@ func (s *SchemaChangeManager) Start(stopper *stop.Stopper) {
 						// A schema change execution might fail soon after
 						// unsetting UpVersion, and we still want to process
 						// outstanding mutations. Similar with a table marked for deletion.
-						if table.UpVersion || table.Deleted() ||
+						if table.UpVersion || table.Deleted() || table.Adding() ||
 							table.Renamed() || len(table.Mutations) > 0 {
 							if log.V(2) {
 								log.Infof(context.TODO(), "%s: queue up pending schema change; table: %d, version: %d",

--- a/sql/sqlbase/structured.go
+++ b/sql/sqlbase/structured.go
@@ -1366,6 +1366,11 @@ func (desc *TableDescriptor) Deleted() bool {
 	return desc.State == TableDescriptor_DROP
 }
 
+// Adding returns true if the table is being added.
+func (desc *TableDescriptor) Adding() bool {
+	return desc.State == TableDescriptor_ADD
+}
+
 // Renamed returns true if the table is being renamed.
 func (desc *TableDescriptor) Renamed() bool {
 	return len(desc.Renames) > 0

--- a/sql/table.go
+++ b/sql/table.go
@@ -146,13 +146,15 @@ func (p *planner) mustGetTableDesc(tn *parser.TableName) (*sqlbase.TableDescript
 }
 
 var errTableDeleted = errors.New("table is being deleted")
+var errTableAdding = errors.New("table is being added")
 
 func filterTableState(tableDesc *sqlbase.TableDescriptor) error {
-	if tableDesc.Deleted() {
+	switch {
+	case tableDesc.Deleted():
 		return errTableDeleted
-	}
-
-	if tableDesc.State != sqlbase.TableDescriptor_PUBLIC {
+	case tableDesc.Adding():
+		return errTableAdding
+	case tableDesc.State != sqlbase.TableDescriptor_PUBLIC:
 		return errors.Errorf("table in unknown state: %s", tableDesc.State.String())
 	}
 	return nil

--- a/sql/table.go
+++ b/sql/table.go
@@ -139,7 +139,23 @@ func (p *planner) mustGetTableDesc(tn *parser.TableName) (*sqlbase.TableDescript
 	if desc == nil {
 		return nil, sqlbase.NewUndefinedTableError(tn.String())
 	}
+	if err := filterTableState(desc); err != nil {
+		return nil, err
+	}
 	return desc, nil
+}
+
+var errTableDeleted = errors.New("table is being deleted")
+
+func filterTableState(tableDesc *sqlbase.TableDescriptor) error {
+	if tableDesc.Deleted() {
+		return errTableDeleted
+	}
+
+	if tableDesc.State != sqlbase.TableDescriptor_PUBLIC {
+		return errors.Errorf("table in unknown state: %s", tableDesc.State.String())
+	}
+	return nil
 }
 
 // getTableLease implements the SchemaAccessor interface.
@@ -159,7 +175,14 @@ func (p *planner) getTableLease(tn *parser.TableName) (*sqlbase.TableDescriptor,
 		//   so they cannot be leased. Instead, we simply return the static
 		//   descriptor and rely on the immutability privileges set on the
 		//   descriptors to cause upper layers to reject mutations statements.
-		return p.mustGetTableDesc(tn)
+		tbl, err := p.mustGetTableDesc(tn)
+		if err != nil {
+			return nil, err
+		}
+		if err := filterTableState(tbl); err != nil {
+			return nil, err
+		}
+		return tbl, nil
 	}
 
 	dbID, err := p.getDatabaseID(tn.Database())
@@ -207,6 +230,17 @@ func (p *planner) getTableLease(tn *parser.TableName) (*sqlbase.TableDescriptor,
 func (p *planner) getTableLeaseByID(tableID sqlbase.ID) (*sqlbase.TableDescriptor, error) {
 	if log.V(2) {
 		log.Infof(p.ctx(), "planner acquiring lease on table ID %d", tableID)
+	}
+
+	if testDisableTableLeases {
+		table, err := sqlbase.GetTableDescFromID(p.txn, tableID)
+		if err != nil {
+			return nil, err
+		}
+		if err := filterTableState(table); err != nil {
+			return nil, err
+		}
+		return table, nil
 	}
 
 	// First, look to see if we already have a lease for this table -- including

--- a/sql/tablewriter.go
+++ b/sql/tablewriter.go
@@ -212,7 +212,7 @@ type tableUpserter struct {
 	// Set by init.
 	txn                   *client.Txn
 	tableDesc             *sqlbase.TableDescriptor
-	fkTables              TablesByID // for fk checks in update case
+	fkTables              tableLookupsByID // for fk checks in update case
 	ru                    rowUpdater
 	updateColIDtoRowIndex map[sqlbase.ColumnID]int
 	a                     sqlbase.DatumAlloc

--- a/sql/tablewriter.go
+++ b/sql/tablewriter.go
@@ -527,6 +527,12 @@ func (td *tableDeleter) fastPathAvailable(ctx context.Context) bool {
 		}
 		return false
 	}
+	if len(td.rd.helper.tableDesc.PrimaryIndex.ReferencedBy) > 0 {
+		if log.V(2) {
+			log.Info(ctx, "delete forced to scan: table is referenced by foreign keys")
+		}
+		return false
+	}
 	return true
 }
 

--- a/sql/testdata/fk
+++ b/sql/testdata/fk
@@ -327,3 +327,20 @@ DELETE FROM pairs WHERE id = 2
 #
 # statement error foreign key violation: value \[2\] not found in other@primary \[x\]
 # INSERT INTO crossdb(y) VALUES (2)
+
+
+# since PKs are handled differently than other indexes, check pk<->pk ref with no other indexes in play.
+statement ok
+CREATE TABLE foo (id INT PRIMARY KEY)
+
+statement ok
+CREATE TABLE bar (id INT PRIMARY KEY REFERENCES foo)
+
+statement ok
+INSERT INTO foo VALUES (2)
+
+statement ok
+INSERT INTO bar VALUES (2)
+
+statement error foreign key violation: value\(s\) \[2] in columns \[id\] referenced in table "bar"
+DELETE FROM foo

--- a/sql/testdata/fk
+++ b/sql/testdata/fk
@@ -66,6 +66,9 @@ CREATE TABLE reviews (
 statement ok
 INSERT INTO orders VALUES (1, '780', 2);
 
+statement error foreign key violation: value\(s\) \['780'\] in columns \[sku\] referenced in table "orders"
+DELETE FROM products
+
 statement ok
 INSERT INTO reviews VALUES (1, '780', 2, 1, NULL)
 
@@ -315,20 +318,6 @@ UPDATE pairs SET dest = 'too' WHERE id = 2
 statement error foreign key violation: value\(s\) \[200 'two'\] in columns \[src dest\] referenced in table "refpairs"
 DELETE FROM pairs WHERE id = 2
 
-# Disabled due to issue #8392.
-# statement ok
-# CREATE DATABASE test2
-#
-# statement ok
-# CREATE TABLE test2.other(x INT PRIMARY KEY)
-#
-# statement ok
-# CREATE TABLE crossdb (y INT PRIMARY KEY, FOREIGN KEY (y) REFERENCES test2.other(x))
-#
-# statement error foreign key violation: value \[2\] not found in other@primary \[x\]
-# INSERT INTO crossdb(y) VALUES (2)
-
-
 # since PKs are handled differently than other indexes, check pk<->pk ref with no other indexes in play.
 statement ok
 CREATE TABLE foo (id INT PRIMARY KEY)
@@ -344,3 +333,24 @@ INSERT INTO bar VALUES (2)
 
 statement error foreign key violation: value\(s\) \[2] in columns \[id\] referenced in table "bar"
 DELETE FROM foo
+
+statement ok
+CREATE DATABASE otherdb
+
+statement ok
+CREATE TABLE otherdb.othertable (id INT PRIMARY KEY)
+
+statement ok
+CREATE TABLE crossdb (id INT PRIMARY KEY, FOREIGN KEY (id) REFERENCES otherdb.othertable)
+
+statement error foreign key violation: value \[2\] not found in othertable@primary \[id\]
+INSERT INTO crossdb VALUES (2)
+
+statement ok
+INSERT INTO otherdb.othertable VALUES (1), (2)
+
+statement ok
+INSERT INTO crossdb VALUES (2)
+
+statement error foreign key violation: value\(s\) \[2] in columns \[id\] referenced in table "crossdb"
+DELETE FROM otherdb.othertable WHERE id = 2

--- a/sql/truncate.go
+++ b/sql/truncate.go
@@ -51,7 +51,7 @@ func (p *planner) Truncate(n *parser.Truncate) (planNode, error) {
 			return nil, err
 		}
 
-		fkTables := TablesNeededForFKs(*tableDesc, CheckDeletes)
+		fkTables := tablesNeededForFKs(*tableDesc, CheckDeletes)
 		if err := p.fillFKTableMap(fkTables); err != nil {
 			return nil, err
 		}

--- a/sql/update.go
+++ b/sql/update.go
@@ -170,7 +170,7 @@ func (p *planner) Update(n *parser.Update, desiredTypes []parser.Datum, autoComm
 		requestedCols = en.tableDesc.Columns
 	}
 
-	fkTables := TablesNeededForFKs(*en.tableDesc, CheckUpdates)
+	fkTables := tablesNeededForFKs(*en.tableDesc, CheckUpdates)
 	if err := p.fillFKTableMap(fkTables); err != nil {
 		return nil, err
 	}

